app/updatemanager.go

```diff
@@ -125,9 +142,143 @@ func (u *UpdateManager) run(ctx context.Context) error {
 	return nil
 }
 
+type UpdateControlMapState struct {
+	Action           string `json:"action"`
+	OnMapExpire      string `json:"on_map_expire"`
+	OnActionExecuted string `json:"on_action_executed"`
+}
+
 type UpdateControlMap struct {
-	ID       string `json:"id"`
-	Priority int    `json:"priority"`
+	ID       string                           `json:"id"`
+	Priority int                              `json:"priority"`
+	States   map[string]UpdateControlMapState `json:"states"`
+}
+
+var UpdateControlMapStateKeyValid []string = []string{
+	"ArtifactInstall_Enter",
+	"ArtifactReboot_Enter",
+	"ArtifactCommit_Enter",
+}
+var UpdateControlMapStateActionValid []string = []string{
+	"continue",
+	"force_continue",
+	"pause",
+	"fail",
+}
+var UpdateControlMapStateOnMapExpireValid []string = []string{
+	"continue",
+	"force_continue",
+	"fail",
+}
+var UpdateControlMapStateOnActionExecutedValid []string = UpdateControlMapStateActionValid
+
+var UpdateControlMapStateActionDefault string = "continue"
+
+func UpdateControlMapStateOnMapExpireDefault(action string) string {
+	if action == "pause" {
+		return "fail"
+	}
+	return action
+}
+
+var UpdateControlMapStateOnActionExecutedDefault = UpdateControlMapStateActionDefault
+
+func (s UpdateControlMapState) Validate() error {
+	if s.Action != "" {
+		found, err := utils.ElemInSlice(UpdateControlMapStateActionValid, s.Action)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("invalid value %q", s.Action)
+		}
+	}
+
+	if s.OnMapExpire != "" {
+		found, err := utils.ElemInSlice(UpdateControlMapStateOnMapExpireValid, s.OnMapExpire)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("invalid value %q", s.OnMapExpire)
+		}
+	}
+
+	if s.OnActionExecuted != "" {
+		found, err := utils.ElemInSlice(UpdateControlMapStateOnActionExecutedValid, s.OnActionExecuted)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("invalid value %q", s.OnActionExecuted)
+		}
+	}
+
+	return nil
+}
+
+func (s *UpdateControlMapState) Sanitize() {
+	if s.Action == "" {
+		log.Debugf("Action was empty, setting to default %q", UpdateControlMapStateActionDefault)
+		s.Action = UpdateControlMapStateActionDefault
+	}
+	if s.OnMapExpire == "" {
+		onMapExpireDefault := UpdateControlMapStateOnMapExpireDefault(s.Action)
+		log.Debugf("OnMapExpire was empty, setting to default %q", onMapExpireDefault)
+		s.OnMapExpire = onMapExpireDefault
+	}
+
+	if s.OnActionExecuted == "" {
+		log.Debugf("OnActionExecuted was empty, setting to default %q", UpdateControlMapStateOnActionExecutedDefault)
+		s.OnActionExecuted = UpdateControlMapStateOnActionExecutedDefault
+	}
+}
+
+func (m UpdateControlMap) Validate() error {
+	// ID is mandatory
+	if m.ID == "" {
+		return errors.New("ID cannot be empty")
+	}
+
+	// Priority must be 0 or higher
+	if m.Priority < 0 {
```

Comment:
Priority can be any number, and there are valid reasons for it to be negative. Since we don't expose priority in the UI, it will always use 0. Therefore, using negative priorities for a local UI process allows a setup where the UI/backend always has the last word. IOW, the device operator has some control over the update progress, unless the central UI/backend operator overrides it.

We could consider limiting its range though. I can hardly imagine why anyone would exceed -10 and 10, so that could be a possibility.

Reply:
@lluiscampos: You haven't forgotten this, have you? :-)

Reply:
No, I haven't!

Reply:
See https://github.com/mendersoftware/mender/pull/767